src/centralized_telescope.rs

```diff
@@ -0,0 +1,421 @@
+}
+impl Setup {
+    fn param_small_case(params: &Params, u_f64: f64, sec_param: u64) -> Self {
+        let ln12 = (12f64).ln();
+        let d = (32.0 * ln12 * u_f64).ceil();
+        Self {
+            sec_param,
+            n_p: params.n_p,
+            u: u_f64 as u64,
+            r: params.lambda_rel as u64,
+            d: d as u64,
+            q: 2.0 * ln12 / d,
+            b: (8.0 * (u_f64 + 1.0) * d / ln12).floor() as u64,
+        }
+    }
+
+    fn param_high_case(params: &Params, u_f64: f64, lambda_rel2: f64, sec_param: u64) -> Self {
+        let loge = f64::from(LOG2_E);
+        let l2 = lambda_rel2 + 2.0;
+        let d = (16.0 * u_f64 * l2 / loge).ceil();
+        debug_assert!(params.n_p as f64 >= d * d * loge / (9.0 * l2));
+        Self {
+            sec_param,
+            n_p: params.n_p,
+            u: u_f64 as u64,
+            r: (params.lambda_rel / lambda_rel2).ceil() as u64,
+            d: d as u64,
+            q: 2.0 * l2 / (d * loge),
+            b: (((l2 + u_f64.log2()) / l2) * (3.0 * u_f64 * d / 4.0) + d + u_f64).floor() as u64,
+        }
+    }
+
+    fn param_mid_case(params: &Params, u_f64: f64, s1: f64, sec_param: u64) -> Self {
+        fn compute_w(u: f64, l: f64) -> f64 {
+            fn factorial_check(w: f64, l: f64) -> bool {
+                let bound = (-l).exp2();
+                let mut factor = (w.ceil() as u64).saturating_add(1);
+                let w_2 = w + 2.0;
+                let exp_1_over_w = E.powf(1.0 / w);
+                let mut ratio = (14.0 * w * w * w_2 * exp_1_over_w) / (w_2 - exp_1_over_w);
+                while factor != 0 {
+                    ratio /= factor as f64;
+                    if ratio <= bound {
+                        return true;
+                    }
+                    factor = factor.saturating_sub(1);
+                }
+                false
+            }
+            let mut w: f64 = u;
+            while !factorial_check(w, l) {
+                w += 1.0;
+            }
+            w
+        }
+        let lambda_rel1 = params.lambda_rel.min(s1);
+        let lbar = (lambda_rel1 + 7.0) / f64::from(LOG2_E);
+        let d = (16.0 * u_f64 * lbar).ceil();
+        let lbar_over_d = lbar / d;
+        debug_assert!(params.n_p as f64 >= d * d / (9.0 * lbar));
+
+        let w = compute_w(u_f64, lambda_rel1);
+        let exponential = (2.0 * u_f64 * w * lbar / params.n_p as f64 + 7.0 * u_f64 / w).exp();
+        Self {
+            sec_param,
+            n_p: params.n_p,
+            u: u_f64 as u64,
+            r: (params.lambda_rel / lambda_rel1).ceil() as u64,
+            d: d as u64,
+            q: 2.0 * lbar_over_d,
+            b: ((w * lbar_over_d + 1.0) * exponential * d * u_f64 + d).floor() as u64,
+        }
+    }
+
+    /// Setup algorithm taking a Params as input and returning setup parameters (u,d,q)
+    pub fn new(params: &Params) -> Self {
+        let sec_param = params.lambda_rel.max(params.lambda_sec).ceil() as u64;
```

Comment:
Another problem, if we use lambda directly it may overflow:
```
        let epsilon_fail: u64 = 1 << sec_param;
```
Perhaps we should indeed just hardcode a value like before.

Reply:
Removed.